File: public/js/supabase-client.js
/**
 * Supabase Client for Pudidi CMG Frontend
 * ========================================
 * Replaces Gist/Cache file fetching with Supabase REST API
 *
 * Uses anon key for public read-only access
 */

const SUPABASE_URL = 'https://btyfbrclgmphcjgrvcgd.supabase.co';
const SUPABASE_ANON_KEY = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6ImJ0eWZicmNsZ21waGNqZ3J2Y2dkIiwicm9sZSI6ImFub24iLCJpYXQiOjE3NjEwMDk1MjksImV4cCI6MjA3NjU4NTUyOX0.WQK2xJMa6YWUABsXq2MQwJGpOQHt5GfZJ5pLe7MZIi8';

class SupabaseAPI {
    constructor() {
        this.baseURL = `${SUPABASE_URL}/rest/v1`;
        this.headers = {
            'apikey': SUPABASE_ANON_KEY,
            'Authorization': `Bearer ${SUPABASE_ANON_KEY}`,
            'Content-Type': 'application/json'
        };
    }

    /**
     * Get CMG Online (actual values)
     * @param {Object} params - Query parameters
     * @param {string} params.startDate - Start date (YYYY-MM-DD)
     * @param {string} params.endDate - End date (YYYY-MM-DD)
     * @param {string} params.node - Node name (optional)
     * @param {number} params.limit - Max records (default 1000)
     */
    async getCMGOnline(params = {}) {
        const {
            startDate,
            endDate,
            node,
            limit = 1000
        } = params;

        const queryParams = new URLSearchParams({
            select: '*',
            order: 'datetime.asc',
            limit: limit.toString()
        });

        if (startDate) {
            queryParams.append('date', `gte.${startDate}`);
        }
        if (endDate) {
            queryParams.append('date', `lte.${endDate}`);
        }
        if (node) {
            queryParams.append('node', `eq.${node}`);
        }

        const url = `${this.baseURL}/cmg_online?${queryParams}`;
        const response = await fetch(url, { headers: this.headers });

        if (!response.ok) {
            throw new Error(`Supabase API error: ${response.statusText}`);
        }

        return await response.json();
    }

    /**
     * Get CMG Programado (official forecasts)
     * @param {Object} params - Query parameters
     * @param {string} params.startDate - Start date (YYYY-MM-DD)
     * @param {string} params.endDate - End date (YYYY-MM-DD)
     * @param {number} params.limit - Max records (default 1000)
     */
    async getCMGProgramado(params = {}) {
        const {
            startDate,
            endDate,
            limit = 1000
        } = params;

        const queryParams = new URLSearchParams({
            select: '*',
            order: 'datetime.asc',
            limit: limit.toString()
        });

        if (startDate) {
            queryParams.append('date', `gte.${startDate}`);
        }
        if (endDate) {
            queryParams.append('date', `lte.${endDate}`);
        }

        const url = `${this.baseURL}/cmg_programado?${queryParams}`;
        const response = await fetch(url, { headers: this.headers });

        if (!response.ok) {
            throw new Error(`Supabase API error: ${response.statusText}`);
        }

        return await response.json();
    }

    /**
     * Get ML Predictions
     * @param {Object} params - Query parameters
     * @param {string} params.forecastDatetime - Specific forecast datetime
     * @param {string} params.startDate - Start date for forecast_datetime
     * @param {string} params.endDate - End date for forecast_datetime
     * @param {number} params.limit - Max records (default 5000)
     */
    async getMLPredictions(params = {}) {
        const {
            forecastDatetime,
            startDate,
            endDate,
            limit = 5000
        } = params;

        const queryParams = new URLSearchParams({
            select: '*',
            order: 'forecast_datetime.desc,target_datetime.asc',
            limit: limit.toString()
        });

        if (forecastDatetime) {
            queryParams.append('forecast_datetime', `eq.${forecastDatetime}`);
        }
        if (startDate) {
            queryParams.append('forecast_datetime', `gte.${startDate}`);
        }
        if (endDate) {
            queryParams.append('forecast_datetime', `lte.${endDate}`);
        }

        const url = `${this.baseURL}/ml_predictions?${queryParams}`;
        const response = await fetch(url, { headers: this.headers });

        if (!response.ok) {
            throw new Error(`Supabase API error: ${response.statusText}`);
        }

        return await response.json();
    }

    /**
     * Get latest ML forecast (most recent 24-hour prediction)
     */
    async getLatestMLForecast() {
        // First get the latest forecast_datetime
        let url = `${this.baseURL}/ml_predictions?select=forecast_datetime&order=forecast_datetime.desc&limit=1`;
        let response = await fetch(url, { headers: this.headers });

        if (!response.ok) {
            throw new Error(`Supabase API error: ${response.statusText}`);
        }

        const latest = await response.json();
        if (latest.length === 0) {
            return [];
        }

        const latestForecastTime = latest[0].forecast_datetime;

        // Now get all predictions from that forecast
        url = `${this.baseURL}/ml_predictions?forecast_datetime=eq.${latestForecastTime}&order=target_datetime.asc`;
        response = await fetch(url, { headers: this.headers });

        if (!response.ok) {
            throw new Error(`Supabase API error: ${response.statusText}`);
        }

        return await response.json();
    }

    /**
     * Get available forecast dates/times for ML predictions
     * Returns unique forecast_datetime values
     */
    async getAvailableMLForecasts() {
        const url = `${this.baseURL}/ml_predictions?select=forecast_datetime&order=forecast_datetime.desc&limit=500`;
        const response = await fetch(url, { headers: this.headers });

        if (!response.ok) {
            throw new Error(`Supabase API error: ${response.statusText}`);
        }

        const data = await response.json();

        // Get unique forecast datetimes
        const uniqueForecasts = [...new Set(data.map(d => d.forecast_datetime))];

        // Group by date and hour
        const grouped = {};
        uniqueForecasts.forEach(dt => {
            const date = dt.split('T')[0];
            const hour = new Date(dt).getUTCHours();

            if (!grouped[date]) {
                grouped[date] = [];
            }
            grouped[date].push({
                datetime: dt,
                hour: hour
            });
        });

        return grouped;
    }

    /**
     * Get latest 24 hours of CMG Online data
     */
    async getLatest24HoursCMG() {
        const url = `${this.baseURL}/cmg_online?select=*&order=datetime.desc&limit=72`;  // 3 nodes × 24 hours
        const response = await fetch(url, { headers: this.headers });

        if (!response.ok) {
            throw new Error(`Supabase API error: ${response.statusText}`);
        }

        return await response.json();
    }

    /**
     * Format Supabase data to match old Gist structure
     * (for compatibility with existing frontend code)
     */
    formatCMGOnlineAsCache(records) {
        return {
            metadata: {
                last_update: new Date().toISOString(),
                total_records: records.length
            },
            data: records.map(r => ({
                datetime: r.datetime,
                date: r.date,
                hour: r.hour,
                node: r.node,
                cmg_usd: r.cmg_usd,
                source: r.source || 'supabase'
            }))
        };
    }

    formatMLPredictionsAsCache(records) {
        // Group by forecast_datetime
        const grouped = {};
        records.forEach(r => {
            const forecastKey = r.forecast_datetime;
            if (!grouped[forecastKey]) {
                grouped[forecastKey] = {
                    forecast_time: r.forecast_datetime,
                    model_version: r.model_version,
                    predictions: []
                };
            }
            grouped[forecastKey].predictions.push({
                horizon: r.horizon,
                target_datetime: r.target_datetime,
                cmg: r.cmg_predicted,
                prob_zero: r.prob_zero,
                threshold: r.threshold
            });
        });

        // Convert to daily_data structure
        const daily_data = {};
        Object.values(grouped).forEach(forecast => {
            const date = forecast.forecast_time.split('T')[0];
            const hour = new Date(forecast.forecast_time).getUTCHours();

            if (!daily_data[date]) {
                daily_data[date] = { ml_forecasts: {} };
            }
            daily_data[date].ml_forecasts[hour.toString()] = forecast;
        });

        return {
            metadata: {
                last_update: new Date().toISOString(),
                total_forecasts: records.length
            },
            daily_data: daily_data
        };
    }
}

// Export for use in other scripts
if (typeof module !== 'undefined' && module.exports) {
    module.exports = SupabaseAPI;
}

// Create global instance
const supabaseAPI = new SupabaseAPI();

File: lib/utils/supabase_client.py
"""
Supabase Client for CMG Prediction System
==========================================

Handles all database operations for:
- CMG Online (historical data)
- CMG Programado (forecast data)
- ML Predictions (24-hour forecasts)

Replaces GitHub Gist storage to avoid rate limiting.
"""

import os
import requests
import json
from datetime import datetime, date
from typing import List, Dict, Any, Optional
import pytz

class SupabaseClient:
    """Client for interacting with Supabase PostgreSQL database"""
    
    def __init__(self):
        """Initialize Supabase client with credentials from environment"""
        self.supabase_url = os.environ.get('SUPABASE_URL')
        self.supabase_key = os.environ.get('SUPABASE_SERVICE_KEY')  # Use service_role key for writes
        
        if not self.supabase_url or not self.supabase_key:
            raise ValueError(
                "Missing Supabase credentials. Set SUPABASE_URL and SUPABASE_SERVICE_KEY environment variables."
            )
        
        self.base_url = f"{self.supabase_url}/rest/v1"
        self.headers = {
            "apikey": self.supabase_key,
            "Authorization": f"Bearer {self.supabase_key}",
            "Content-Type": "application/json",
            "Prefer": "return=minimal"  # Don't return inserted data by default
        }
    
    # ========================================
    # CMG ONLINE (HISTORICAL DATA)
    # ========================================
    
    def insert_cmg_online_batch(self, records: List[Dict[str, Any]]) -> bool:
        """
        Insert batch of CMG Online records.
        Uses UPSERT to handle duplicates gracefully.

        Args:
            records: List of dicts with keys: datetime, date, hour, node, cmg_usd

        Returns:
            True if successful, False otherwise
        """
        if not records:
            print("⚠️  No records to insert")
            return True

        try:
            # Use on_conflict query parameter to specify UPSERT behavior
            # This tells PostgREST to update existing records instead of failing
            url = f"{self.base_url}/cmg_online?on_conflict=datetime,node"
            headers = self.headers.copy()
            headers["Prefer"] = "resolution=merge-duplicates,return=minimal"

            response = requests.post(url, json=records, headers=headers)

            if response.status_code in [200, 201, 204]:
                print(f"✅ Inserted {len(records)} CMG Online records")
                return True
            else:
                print(f"❌ Failed to insert CMG Online: {response.status_code}")
                print(f"   Response: {response.text}")
                # Still return True if it's just duplicate errors (script should continue)
                if response.status_code == 409:
                    print("   Note: Some records already exist (this is OK)")
                    return True
                return False
        except Exception as e:
            print(f"❌ Error inserting CMG Online: {e}")
            return False
    
    def get_cmg_online(
        self, 
        start_date: Optional[str] = None,
        end_date: Optional[str] = None,
        node: Optional[str] = None,
        limit: int = 1000
    ) -> List[Dict[str, Any]]:
        """
        Get CMG Online records with optional filters.
        
        Args:
            start_date: Filter by date >= this (YYYY-MM-DD)
            end_date: Filter by date <= this (YYYY-MM-DD)
            node: Filter by specific node
            limit: Max records to return
        
        Returns:
            List of CMG Online records
        """
        try:
            url = f"{self.base_url}/cmg_online"

            # Build params as list of tuples to allow multiple values for same key
            # This is required for PostgREST to AND multiple filters on the same column
            params = [("order", "datetime.desc"), ("limit", limit)]

            # Handle date range filters properly (PostgREST syntax)
            if start_date:
                params.append(("date", f"gte.{start_date}"))
            if end_date:
                params.append(("date", f"lte.{end_date}"))

            if node:
                params.append(("node", f"eq.{node}"))

            response = requests.get(url, params=params, headers=self.headers)
            
            if response.status_code == 200:
                return response.json()
            else:
                print(f"❌ Failed to get CMG Online: {response.status_code}")
                return []
        except Exception as e:
            print(f"❌ Error getting CMG Online: {e}")
            return []
    
    # ========================================
    # CMG PROGRAMADO (FORECAST DATA)
    # ========================================
    
    def insert_cmg_programado_batch(self, records: List[Dict[str, Any]]) -> bool:
        """
        Insert batch of CMG Programado records.
        Uses UPSERT to handle duplicates.

        Args:
            records: List of dicts with keys:
                - forecast_datetime: When forecast was made
                - forecast_date: Date when forecast was made
                - forecast_hour: Hour when forecast was made
                - target_datetime: What hour is being predicted
                - target_date: Date being predicted
                - target_hour: Hour being predicted
                - node: Node name
                - node_id: Node ID (required after migration 002)
                - cmg_usd: CMG value in USD

        Returns:
            True if successful
        """
        try:
            # FIXED: Use node_id instead of node to match unique constraint
            # unique_cmg_prog_forecast_target_node_id (forecast_datetime, target_datetime, node_id)
            url = f"{self.base_url}/cmg_programado?on_conflict=forecast_datetime,target_datetime,node_id"
            headers = self.headers.copy()
            headers["Prefer"] = "resolution=merge-duplicates,return=minimal"

            response = requests.post(url, json=records, headers=headers)

            if response.status_code in [200, 201, 204]:
                print(f"✅ Inserted {len(records)} CMG Programado records")
                return True
            else:
                print(f"❌ Failed to insert CMG Programado: {response.status_code}")
                print(f"   Response: {response.text}")
                # Still return True if it's just duplicate errors (script should continue)
                if response.status_code == 409:
                    print("   Note: Some records already exist (this is OK)")
                    return True
                return False
        except Exception as e:
            print(f"❌ Error inserting CMG Programado: {e}")
            return False

    def get_cmg_programado(
        self,
        start_date: Optional[str] = None,
        end_date: Optional[str] = None,
        node: Optional[str] = None,
        limit: int = 10000,
        latest_forecast_only: bool = True
    ) -> List[Dict[str, Any]]:
        """
        Get CMG Programado forecast records.

        Args:
            start_date: Filter by target_date >= this (YYYY-MM-DD)
            end_date: Filter by target_date <= this (YYYY-MM-DD)
            node: Filter by specific node
            limit: Max records to return
            latest_forecast_only: If True, only return records from the latest forecast (prevents duplicates)

        Returns:
            List of forecast records with forecast_datetime and target_datetime
        """
        try:
            url = f"{self.base_url}/cmg_programado"

            if latest_forecast_only:
                # STEP 1: Get the latest forecast_datetime for this node
                params_latest = [
                    ("select", "forecast_datetime"),
                    ("order", "forecast_datetime.desc"),
                    ("limit", "1")
                ]
                if node:
                    params_latest.append(("node", f"eq.{node}"))

                response_latest = requests.get(url, params=params_latest, headers=self.headers)

                if response_latest.status_code != 200 or not response_latest.json():
                    print("⚠️ No forecasts found")
                    return []

                latest_forecast_dt = response_latest.json()[0]['forecast_datetime']

                # STEP 2: Get all records from this specific forecast
                params = [
                    ("forecast_datetime", f"eq.{latest_forecast_dt}"),
                    ("order", "target_datetime.asc"),
                    ("limit", limit)
                ]
            else:
                # Original behavior: get multiple forecasts (may contain duplicates)
                params = [("order", "forecast_datetime.desc,target_datetime.asc"), ("limit", limit)]

            # Handle date range filters properly (PostgREST syntax)
            if start_date:
                params.append(("target_date", f"gte.{start_date}"))
            if end_date:
                params.append(("target_date", f"lte.{end_date}"))
            if node:
                params.append(("node", f"eq.{node}"))

            response = requests.get(url, params=params, headers=self.headers)

            if response.status_code == 200:
                return response.json()
            else:
                print(f"❌ Failed to get CMG Programado: {response.status_code}")
                return []
        except Exception as e:
            print(f"❌ Error getting CMG Programado: {e}")
            return []
    
    # ========================================
    # ML PREDICTIONS (24H FORECAST)
    # ========================================
    
    def insert_ml_predictions_batch(self, records: List[Dict[str, Any]]) -> bool:
        """
        Insert batch of ML prediction records.
        Uses UPSERT to handle duplicates (when same forecast is run multiple times).

        Args:
            records: List of dicts with keys:
                     forecast_datetime, target_datetime, horizon,
                     cmg_predicted, prob_zero, threshold, model_version

        Returns:
            True if successful
        """
        try:
            # FIXED: Add on_conflict to match unique constraint
            # ml_predictions_forecast_datetime_target_datetime_key (forecast_datetime, target_datetime)
            url = f"{self.base_url}/ml_predictions?on_conflict=forecast_datetime,target_datetime"
            headers = self.headers.copy()
            headers["Prefer"] = "resolution=merge-duplicates,return=minimal"

            response = requests.post(url, json=records, headers=headers)

            if response.status_code in [200, 201, 204]:
                print(f"✅ Inserted {len(records)} ML prediction records")
                return True
            else:
                print(f"❌ Failed to insert ML predictions: {response.status_code}")
                print(f"   Response: {response.text}")
                # Still return True if it's just duplicate errors (script should continue)
                if response.status_code == 409:
                    print("   Note: Some predictions already exist (this is OK)")
                    return True
                return False
        except Exception as e:
            print(f"❌ Error inserting ML predictions: {e}")
            return False
    
    def get_nodes(self) -> List[Dict[str, Any]]:
        """
        Get all nodes from the nodes table.

        Returns:
            List of node records with id, code, name, etc.
        """
        try:
            url = f"{self.base_url}/nodes"
            params = {"order": "code.asc"}

            response = requests.get(url, params=params, headers=self.headers)

            if response.status_code == 200:
                return response.json()
            else:
                print(f"❌ Failed to get nodes: {response.status_code}")
                return []
        except Exception as e:
            print(f"❌ Error getting nodes: {e}")
            return []

    def get_node_id_map(self) -> Dict[str, int]:
        """
        Get mapping of node code -> node_id for efficient lookups.

        Returns:
            Dict mapping node codes to their IDs
        """
        nodes = self.get_nodes()
        return {node['code']: node['id'] for node in nodes}

    def get_latest_ml_predictions(self, limit: int = 24) -> List[Dict[str, Any]]:
        """
        Get the most recent ML forecast (latest 24 predictions).

        Returns:
            List of ML prediction records from most recent forecast
        """
        try:
            # First, get the latest forecast_datetime
            url = f"{self.base_url}/ml_predictions"
            params = {
                "select": "forecast_datetime",
                "order": "forecast_datetime.desc",
                "limit": 1
            }

            response = requests.get(url, params=params, headers=self.headers)

            if response.status_code != 200 or not response.json():
                print("❌ No ML predictions found")
                return []

            latest_forecast = response.json()[0]["forecast_datetime"]

            # Now get all predictions from that forecast
            params = {
                "forecast_datetime": f"eq.{latest_forecast}",
                "order": "target_datetime.asc",
                "limit": limit
            }

            response = requests.get(url, params=params, headers=self.headers)

            if response.status_code == 200:
                return response.json()
            else:
                print(f"❌ Failed to get ML predictions: {response.status_code}")
                return []
        except Exception as e:
            print(f"❌ Error getting ML predictions: {e}")
            return []

    def get_ml_predictions(
        self,
        start_date: Optional[str] = None,
        end_date: Optional[str] = None,
        limit: int = 10000
    ) -> List[Dict[str, Any]]:
        """
        Get ML predictions with date range filtering.

        Args:
            start_date: Filter by target_datetime >= this (YYYY-MM-DD)
            end_date: Filter by target_datetime <= this (YYYY-MM-DD)
            limit: Max records to return

        Returns:
            List of ML prediction records
        """
        try:
            url = f"{self.base_url}/ml_predictions"

            # Build params as list of tuples to allow multiple values for same key
            params = [
                ("order", "forecast_datetime.desc,target_datetime.asc"),
                ("limit", limit)
            ]

            # Handle date range filters properly (PostgREST syntax)
            if start_date:
                params.append(("target_datetime", f"gte.{start_date}"))
            if end_date:
                params.append(("target_datetime", f"lte.{end_date}"))

            response = requests.get(url, params=params, headers=self.headers)

            if response.status_code == 200:
                return response.json()
            else:
                print(f"❌ Failed to get ML predictions: {response.status_code}")
                return []
        except Exception as e:
            print(f"❌ Error getting ML predictions: {e}")
            return []

    # ========================================
    # METADATA
    # ========================================
    
    def update_metadata(self, key: str, value: Dict[str, Any]) -> bool:
        """Update system metadata"""
        try:
            url = f"{self.base_url}/system_metadata"
            params = {"key": f"eq.{key}"}
            data = {
                "value": value,
                "updated_at": datetime.now(pytz.UTC).isoformat()
            }
            
            response = requests.patch(url, params=params, json=data, headers=self.headers)
            
            if response.status_code in [200, 201, 204]:
                print(f"✅ Updated metadata: {key}")
                return True
            else:
                print(f"❌ Failed to update metadata: {response.status_code}")
                return False
        except Exception as e:
            print(f"❌ Error updating metadata: {e}")
            return False


    # ========================================
    # FORMAT HELPERS (for API compatibility)
    # ========================================

    def format_cmg_online_as_cache(self, records: List[Dict[str, Any]]) -> Dict[str, Any]:
        """
        Format Supabase CMG Online records to match cache file structure.
        Used by API endpoints for backward compatibility.
        """
        daily_data = {}
        santiago_tz = pytz.timezone('America/Santiago')

        for record in records:
            date_str = record['date'] if isinstance(record['date'], str) else str(record['date'])
            node = record['node']
            hour = record['hour']

            if date_str not in daily_data:
                daily_data[date_str] = {
                    'hours': [],
                    'cmg_online': {}
                }

            if node not in daily_data[date_str]['cmg_online']:
                daily_data[date_str]['cmg_online'][node] = {
                    'cmg_usd': [None] * 24
                }

            daily_data[date_str]['cmg_online'][node]['cmg_usd'][hour] = float(record['cmg_usd'])

            if hour not in daily_data[date_str]['hours']:
                daily_data[date_str]['hours'].append(hour)

        # Sort hours
        for date_data in daily_data.values():
            date_data['hours'].sort()

        return {
            'metadata': {
                'last_update': datetime.now(santiago_tz).isoformat(),
                'total_records': len(records),
                'source': 'supabase'
            },
            'daily_data': daily_data
        }

    def format_cmg_programado_as_cache(self, records: List[Dict[str, Any]]) -> Dict[str, Any]:
        """
        Format Supabase CMG Programado records to match cache file structure.
        """
        daily_data = {}
        santiago_tz = pytz.timezone('America/Santiago')

        for record in records:
            date_str = record['date'] if isinstance(record['date'], str) else str(record['date'])
            hour = record['hour']

            if date_str not in daily_data:
                daily_data[date_str] = {
                    'cmg_programado': [None] * 24
                }

            daily_data[date_str]['cmg_programado'][hour] = float(record['cmg_programmed'])

        return {
            'metadata': {
                'last_update': datetime.now(santiago_tz).isoformat(),
                'total_records': len(records),
                'source': 'supabase'
            },
            'daily_data': daily_data
        }


def get_supabase_client() -> SupabaseClient:
    """Get initialized Supabase client"""
    return SupabaseClient()
